### PR TITLE
[jit] get get/setstate working in new format

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23308 [jit] get get/setstate working in new format**
* #23098 [wip][jit] serialize modules as classes
* #23154 [jit] Put all modules in the global Python CU
* #23170 [jit] Make optimize a thread_local flag

This needs to be squashed with the previous PR and landed together, but
I'm pulling it out for review

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23308

Differential Revision: [D16460632](https://our.internmc.facebook.com/intern/diff/D16460632)

### CODE REVIEW

File: torch/csrc/jit/script/python_sugared_value.cpp [hunk 1]
```diff
@@ -239,6 +239,10 @@ std::shared_ptr<SugaredValue> OverloadedMethodValue::call(
                          << err.str();
 }
 
+Value* ModuleValue::asValue(const SourceRange& loc, Function& m) {
```
Comment (suo) [inline torch/csrc/jit/script/python_sugared_value.cpp]: @zdevito do you think this is safe to add now that modules are first-class?

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -691,11 +691,56 @@ struct to_ir {
...
+  // output type of __getstate__. This is necessary so that we can allow
+  // submodules to appear in `state`.
+  std::vector<Argument> emitSetStateArguments(
+      const Def& def,
+      const Self* self,
+      const FunctionSchema& schema,
+      Block* block) {
+    std::vector<Argument> arguments; // for schema
+    TORCH_INTERNAL_ASSERT(def.name().name() == "__setstate__");
+    TORCH_INTERNAL_ASSERT(def.decl().params().size() == 2);
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: This is not an internal  assert?

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -691,11 +691,56 @@ struct to_ir {
...
+    TORCH_INTERNAL_ASSERT(self);
+    {
+      const auto self_param = def.decl().params()[0];
+      const auto& name = self_param.ident().name();
+      Value* new_input = block->addInput()->setDebugName(name);
+      environment_stack->setSugaredVar(
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: I am not a fan of the code duplication here, in particular the environment stack handling. This function should just be providing a type to the argument of  __getstate__ not also duplciating the logic that puts self as a sugared var and initializes the environment. If either of these things change, someone will forget to edit this code.

File: torch/csrc/jit/script/python_sugared_value.cpp (hunk 1 above)
Comment (zdevito) [reply torch/csrc/jit/script/python_sugared_value.cpp]: Yes, it should be there. I am surprised that stuff worked without it actually.

**Review by zdevito — APPROVED** [review]
Looks good, minor nit about code duplication.

File: torch/csrc/jit/script/compiler.cpp (hunk 3 above)
Comment (suo) [reply torch/csrc/jit/script/compiler.cpp]: makes sense

**Comment by suo** [comment]
closing as this has been squashed with https://github.com/pytorch/pytorch/pull/23098

**Pull request closed without merging**